### PR TITLE
Device support for the xiaomi smart wifi socket added

### PR DESCRIPTION
**Pull request by syssi**

### CODE REVIEW

File: mirobo/plug.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+from .device import Device, DeviceException
+
+PROPERTY_POWER = 'power'
+PROPERTY_TEMPERATURE = 'temperature'
+PROPERTY_CURRENT = 'current'
+
+class Plug(Device):
+    """Main class representing the smart wifi socket / plug."""
+
+    def __init__(self, ip: str, token: str, start_id: int = 0, debug: int = 0) -> None:
```
Comment (syssi) [inline mirobo/plug.py]: I don't want to fix this. ;-)

Comment (rytilahti) [inline mirobo/plug.py]: Fair enough, I think hound should complain first after 100 characters..

File: mirobo/plug.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+PROPERTY_TEMPERATURE = 'temperature'
+PROPERTY_CURRENT = 'current'
+
+
+class Plug(Device):
+    """Main class representing the smart wifi socket / plug."""
+
+    def __init__(self, ip: str, token: str, start_id: int = 0, debug: int = 0) -> None:
+        super().__init__(ip, token, start_id, debug)
+        self.manual_seqnum = -1
```
Comment (rytilahti) [inline mirobo/plug.py]: Is the `__init__` here unnecessary though, it just sets the `manual_seqnum` it never uses :-)

File: mirobo/plug.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        """Retrieve properties."""
+        status = self.send(
+            "get_prop",
+            [PROPERTY_POWER, PROPERTY_TEMPERATURE, PROPERTY_CURRENT]
+        )
+        return {
+            PROPERTY_POWER: status[0],
+            PROPERTY_TEMPERATURE: status[1],
+            PROPERTY_CURRENT: status[2]
+        }
```
Comment (rytilahti) [inline mirobo/plug.py]: What type of data is returned by these? Would it make sense to wrap it inside a class like PlugStatus?

File: mirobo/plug.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    def raw_command(self, cmd, params):
+        """Send a raw command to the device."""
+        return self.send(cmd, params)
```
Comment (rytilahti) [inline mirobo/plug.py]: Is `raw_command` necessary when `send()` could be used directly?

**Comment by syssi** [comment]
> Looks good! Have you thought how this could look from the side of the console client?

I have no idea honestly. Just provide a second CLI called "plug"?

File: mirobo/containers.py [hunk 5]
```diff
@@ -39,6 +39,34 @@ def pretty_time(x: int) -> datetime:
...
+
+    @property
+    def current(self) -> float:
+        return self.data["current"]
+
+    def __str__(self) -> str:
+        s = "<PlugStatus power=%s, " % (self.state)
+        s += "temperature=%s, " % (self.temperature)
+        s += "current=%s>" % (self.current)
+        return s
```
Comment (rytilahti) [inline mirobo/containers.py]: I think:
```
s = "<PlugStatus power=%s, temperature=%s, current=%s>" % (self.state,
self.temperature,
self.current)
```
Properly intended would look nicer, but this is fine too.

**Comment by rytilahti** [comment]
That, or using the same client but providing "methods" depending on the discovered device type, considering that there are quite a bit of devices which could be easily supported. I think this can be merged already now.

**Comment by rytilahti** [comment]
Thanks, maybe in the future it makes sense to just extend one tool to cover all the devices, but the cli solution shall work for now :-)

**Comment by syssi** [comment]
```
$ .local/bin/mirobo --ip 192.168.130.57 --token b051cdxxxxxxxxxxxxxxxxxxxxxxxdb -d info
INFO:mirobo.vacuum_cli:Debug mode active
DEBUG:mirobo.vacuum_cli:Read stored sequence ids: {'seq': 7, 'manual_seq': 0}
DEBUG:mirobo.vacuum_cli:Connecting to 192.168.130.57 with token b051cdxxxxxxxxxxxxxxxxxxxxxxxdb
DEBUG:mirobo.protocol:Unable to decrypt, returning raw bytes.
DEBUG:mirobo.device:Discovered 843 14557 with ts: 1970-01-01 01:14:27
DEBUG:mirobo.device:192.168.130.57:54321 >>: {'id': 8, 'params': [], 'method': 'miIO.info'}
DEBUG:mirobo.device:192.168.130.57:54321 (ts: 1970-01-01 01:14:27, id: 8) << {'id': 8, 'result': {'life': 4467, 'hw_ver': 'MW300', 'netif': {'localIp': '192.168.130.57', 'mask': '255.255.255.0', 'gw': '192.168.130.1'}, 'model': 'chuangmi.plug.m1', 'cfg_time': 0, 'otu_stat': [303, 300, 12, 0, 12, 567], 'wifi_fw_ver': 'SD878x-14.76.36.p84-702.1.0-WM', 'ot': 'otu', 'token': 'b051cdxxxxxxxxxxxxxxxxxxxxxxxdb', 'fw_ver': '1.2.4_16', 'mac': '28:6C:FF:FF:FF:FF', 'ap': {'bssid': 'C0:25:06:A8:9A:5E', 'rssi': -69, 'ssid': 'Greendale'}, 'ott_stat': [0, 0, 0, 0], 'mmfree': 30296}}
chuangmi.plug.m1 v1.2.4_16 (28:6C:FF:FF:FF:FF) @ 192.168.130.57 - token: b051cdxxxxxxxxxxxxxxxxxxxxxxxdb
DEBUG:mirobo.vacuum_cli:Full response: {'ap': {'bssid': 'C0:25:06:A8:9A:5E', 'rssi': -69, 'ssid': 'Greendale'},
 'cfg_time': 0,
 'fw_ver': '1.2.4_16',
 'hw_ver': 'MW300',
 'life': 4467,
 'mac': '28:6C:FF:FF:FF:FF',
 'mmfree': 30296,
 'model': 'chuangmi.plug.m1',
 'netif': {'gw': '192.168.130.1',
           'localIp': '192.168.130.57',
           'mask': '255.255.255.0'},
 'ot': 'otu',
 'ott_stat': [0, 0, 0, 0],
 'otu_stat': [303, 300, 12, 0, 12, 567],
 'token': 'b051cdxxxxxxxxxxxxxxxxxxxxxxxdb',
 'wifi_fw_ver': 'SD878x-14.76.36.p84-702.1.0-WM'}
DEBUG:mirobo.vacuum_cli:Writing {'seq': 8, 'manual_seq': 0} to /tmp/python-mirobo.seq
```

**Pull request merged**